// @agentrhq/webcmd/dist/src/help.js
import yaml from 'js-yaml';
import { CLI_COMMAND } from './brand.js';
import { classifyAdapterDomain, commandHelpData as buildCommandHelpData, formatBrowserCommonOptionsHelp, formatCommandHelp, formatCommandListTerm as buildCommandListTerm, formatCommonOptionsHelp, formatSiteCommandDescription as buildSiteCommandDescription, formatSiteHelp, formatRootHelp, siteHelpData as buildSiteHelpData, toPresentableCommand, wrapCommaList as wrapPresentationCommaList, } from './command-presentation.js';
function normalizeStructuredHelpFormat(value) {
    const normalized = value?.toLowerCase();
    if (normalized === 'yaml' || normalized === 'yml')
        return 'yaml';
    if (normalized === 'json')
        return 'json';
    return undefined;
}
export function getRequestedHelpFormat(argv = process.argv) {
    for (let i = 0; i < argv.length; i++) {
        const token = argv[i];
        if (token === '-f' || token === '--format') {
            return normalizeStructuredHelpFormat(argv[i + 1]);
        }
        if (token.startsWith('--format=')) {
            return normalizeStructuredHelpFormat(token.slice('--format='.length));
        }
        if (token.startsWith('-f') && token.length > 2) {
            return normalizeStructuredHelpFormat(token.slice(2));
        }
    }
    return undefined;
}
export function renderStructuredHelp(data, format) {
    if (format === 'json')
        return `${JSON.stringify(data, null, 2)}\n`;
    return yaml.dump(data, { sortKeys: false, lineWidth: 120, noRefs: true });
}
export function wrapCommaList(items, opts = {}) {
    return wrapPresentationCommaList(items, opts);
}
export function classifyAdapter(domain) {
    return classifyAdapterDomain(domain);
}
const installedRootHelpPresentations = new WeakMap();
export function buildRootHelpPresentation(program, groups) {
    const adapterNames = new Set([...groups.external.map((cli) => cli.name), ...groups.apps, ...groups.sites]);
    const commanderHelp = program.createHelp();
    const options = commanderHelp.visibleOptions(program).map((option) => ({
        flags: option.flags,
        description: option.description,
    }));
    return {
        description: program.description(),
        usage: [`${program.name()} [options] [command]`],
        baseText: commanderHelp.formatHelp(program, commanderHelp),
        options,
        commands: program.commands
            .filter((command) => !adapterNames.has(command.name()))
            .map((command) => ({ name: command.name(), description: command.description() })),
        groups: [
            { label: `External CLIs (${groups.external.length})`, items: groups.external.map((cli) => cli.label) },
            { label: `App adapters (${groups.apps.length})`, items: groups.apps },
            { label: `Site adapters (${groups.sites.length})`, items: groups.sites },
        ],
        footer: [
            `Run '${CLI_COMMAND} list' for full command details, or '${CLI_COMMAND} <site> --help' to inspect one site.`,
            `Agent tip: use '${CLI_COMMAND} <site> --help -f yaml' for all command args/options in one structured response.`,
        ],
    };
}
export function getInstalledRootHelpPresentation(command) {
    return installedRootHelpPresentations.get(command);
}
export function installRootPresentationHelp(command, data, presentation) {
    installedRootHelpPresentations.set(command, presentation);
    command.helpInformation = (() => {
        const format = getRequestedHelpFormat();
        if (format)
            return renderStructuredHelp(data(), format);
        return formatRootHelp(presentation);
    });
}
function formatGroupSection(label, names) {
    if (names.length === 0)
        return [];
    return [
        `${label} (${names.length}):`,
        wrapCommaList(names),
        '',
    ];
}
export function formatRootAdapterHelpText(groups) {
    const total = groups.external.length + groups.apps.length + groups.sites.length;
    if (total === 0)
        return '';
    const lines = [''];
    lines.push(...formatGroupSection('External CLIs', groups.external.map(cli => cli.label)));
    lines.push(...formatGroupSection('App adapters', groups.apps));
    lines.push(...formatGroupSection('Site adapters', groups.sites));
    lines.push(`Run '${CLI_COMMAND} list' for full command details, or '${CLI_COMMAND} <site> --help' to inspect one site.`);
    lines.push(`Agent tip: use '${CLI_COMMAND} <site> --help -f yaml' for all command args/options in one structured response.`);
    lines.push('');
    return lines.join('\n');
}
function compactCommanderArgument(arg) {
    return {
        name: arg.name(),
        ...(arg.required ? { required: true } : {}),
        ...(arg.variadic ? { variadic: true } : {}),
        ...(arg.description ? { help: arg.description } : {}),
        ...(arg.defaultValue !== undefined ? { default: arg.defaultValue } : {}),
        ...(arg.argChoices?.length ? { choices: [...arg.argChoices] } : {}),
    };
}
function compactCommanderOption(option) {
    if (option.hidden)
        return null;
    return {
        name: option.attributeName(),
        flags: option.flags,
        ...(option.description ? { help: option.description } : {}),
        ...(option.required ? { takes_value: 'required' } : {}),
        ...(option.optional ? { takes_value: 'optional' } : {}),
        ...(option.mandatory ? { required: true } : {}),
        ...(option.defaultValue !== undefined ? { default: option.defaultValue } : {}),
        ...(option.argChoices?.length ? { choices: [...option.argChoices] } : {}),
        ...(option.negate ? { negate: true } : {}),
    };
}
function compactCommanderOptions(options) {
    return options
        .map(compactCommanderOption)
        .filter((option) => option !== null);
}
/**
 * Extracts a positional placeholder that should appear immediately after this
 * command's name in user-facing path strings. Reads the leading positional
 * (e.g. `<session>`) from a `.usage()` override; commands without a positional
 * override return `null` so the path stays as-is.
 *
 * Example: `browser` declares `.usage('<session> <command> [options]')`,
 * so `commanderPath(browserClickCmd)` becomes
 * `['webcmd', 'browser', '<session>', 'click']`.
 */
export function leadingPositionalFromUsage(command) {
    const usage = command._usage;
    if (!usage)
        return null;
    const match = usage.match(/^\s*(<[^>]+>)/);
    return match ? match[1] : null;
}
function commanderPath(command) {
    const parts = [];
    let current = command;
    while (current) {
        const name = current.name();
        if (name) {
            parts.push(name);
            // If this command declares a leading-positional usage override AND we
            // have already collected a child name below it, the positional must
            // appear between this command and the child (i.e. before the names
            // already collected). parts is in reverse order, so push to the end.
            const positional = leadingPositionalFromUsage(current);
            if (positional && parts.length > 1) {
                // We collected child names first (reverse order). Move them up by one
                // and put the positional at index `parts.length - 2` so reverse()
                // places it between this command and the first child name.
                parts.splice(parts.length - 1, 0, positional);
            }
        }
        current = current.parent;
    }
    return parts.reverse();
}
function commandPathFromRoot(namespaceRoot, command) {
    const rootPath = commanderPath(namespaceRoot);
    const commandPath = commanderPath(command);
    // Strip placeholder positional segments (e.g. `<session>`) from the relative
    // name so agents can still address subcommands by their leaf name. Display
    // paths in `command` / `usage` still include the placeholders.
    return commandPath.slice(rootPath.length).filter(part => !/^<.+>$/.test(part));
}
function collectLeafCommands(command) {
    if (command.commands.length === 0)
        return [command];
    return command.commands.flatMap(child => collectLeafCommands(child));
}
function collectDescendantCommands(command) {
    return command.commands.flatMap(child => [child, ...collectDescendantCommands(child)]);
}
function formatCommanderPositionals(args) {
    return args
        .map(arg => {
        const name = `${arg.name()}${arg.variadic ? '...' : ''}`;
        return arg.required ? `<${name}>` : `[${name}]`;
    })
        .join(' ');
}
function formatCommanderUsage(command, opts = {}) {
    const path = commanderPath(command).join(' ');
    const positionalText = formatCommanderPositionals(command.registeredArguments);
    const hasOptions = compactCommanderOptions(command.options).length > 0
        || (opts.namespaceRoot ? compactCommanderOptions(opts.namespaceRoot.options).length > 0 : false)
        || (opts.globalCommand ? compactCommanderOptions(opts.globalCommand.options).length > 0 : false);
    const optionText = hasOptions ? ' [options]' : '';
    return `${path}${positionalText ? ` ${positionalText}` : ''}${optionText}`;
}
function compactCommanderCommand(namespaceRoot, command, opts = {}) {
    const relativePath = commandPathFromRoot(namespaceRoot, command);
    return {
        name: relativePath.join(' '),
        command: commanderPath(command).join(' '),
        usage: formatCommanderUsage(command, { namespaceRoot, globalCommand: opts.globalCommand }),
        description: command.description(),
        ...(command.aliases().length ? { aliases: command.aliases() } : {}),
        positionals: command.registeredArguments.map(compactCommanderArgument),
        command_options: compactCommanderOptions(command.options),
    };
}
export function commanderNamespaceHelpData(namespaceRoot, opts = {}) {
    const leaves = collectLeafCommands(namespaceRoot)
        .filter(command => command !== namespaceRoot)
        .sort((a, b) => commandPathFromRoot(namespaceRoot, a).join(' ').localeCompare(commandPathFromRoot(namespaceRoot, b).join(' ')));
    // Respect commander's `.usage()` override (e.g. `<session> <command> [options]`
    // on `browser`); fall back to the generic `<command> [args] [options]` form.
    // Read the private `_usage` field directly because `.usage()` returns the
    // auto-generated form if no override was set.
    const commandPath = commanderPath(namespaceRoot).join(' ');
    const usageOverride = namespaceRoot._usage;
    const usage = usageOverride
        ? `${commandPath} ${usageOverride}`
        : `${commandPath} <command> [args] [options]`;
    return {
        namespace: namespaceRoot.name(),
        command: commandPath,
        usage,
        description: opts.description ?? namespaceRoot.description(),
        command_count: leaves.length,
        commands: leaves.map(command => compactCommanderCommand(namespaceRoot, command, opts)),
        namespace_options: compactCommanderOptions(namespaceRoot.options),
        ...(opts.globalCommand ? { global_options: compactCommanderOptions(opts.globalCommand.options) } : {}),
        structured_help: {
            formats: ['yaml', 'json'],
            usage: `${commandPath} --help -f yaml`,
        },
    };
}
export function commanderCommandHelpData(namespaceRoot, command, opts = {}) {
    return {
        namespace: namespaceRoot.name(),
        ...compactCommanderCommand(namespaceRoot, command, opts),
        namespace_options: compactCommanderOptions(namespaceRoot.options),
        ...(opts.globalCommand ? { global_options: compactCommanderOptions(opts.globalCommand.options) } : {}),
        structured_help: {
            formats: ['yaml', 'json'],
            usage: `${commanderPath(command).join(' ')} --help -f yaml`,
        },
    };
}
export function commanderGroupHelpData(namespaceRoot, groupCommand, opts = {}) {
    const leaves = collectLeafCommands(groupCommand)
        .filter(command => command !== groupCommand)
        .sort((a, b) => commandPathFromRoot(namespaceRoot, a).join(' ').localeCompare(commandPathFromRoot(namespaceRoot, b).join(' ')));
    return {
        namespace: namespaceRoot.name(),
        group: commandPathFromRoot(namespaceRoot, groupCommand).join(' '),
        command: commanderPath(groupCommand).join(' '),
        usage: `${commanderPath(groupCommand).join(' ')} <command> [args] [options]`,
        description: groupCommand.description(),
        command_count: leaves.length,
        commands: leaves.map(command => compactCommanderCommand(namespaceRoot, command, opts)),
        namespace_options: compactCommanderOptions(namespaceRoot.options),
        ...(opts.globalCommand ? { global_options: compactCommanderOptions(opts.globalCommand.options) } : {}),
        structured_help: {
            formats: ['yaml', 'json'],
            usage: `${commanderPath(groupCommand).join(' ')} --help -f yaml`,
        },
    };
}
export function installCommanderNamespaceStructuredHelp(namespaceRoot, opts = {}) {
    installStructuredHelp(namespaceRoot, () => commanderNamespaceHelpData(namespaceRoot, opts));
    for (const command of collectDescendantCommands(namespaceRoot)) {
        if (command.commands.length > 0) {
            installStructuredHelp(command, () => commanderGroupHelpData(namespaceRoot, command, opts));
        }
        else {
            installStructuredHelp(command, () => commanderCommandHelpData(namespaceRoot, command, opts));
        }
    }
}
export function formatCommandListTerm(cmd) {
    return buildCommandListTerm(toPresentableCommand(cmd));
}
export function rootHelpData(program, groups) {
    const adapterNames = new Set([...groups.external.map(cli => cli.name), ...groups.apps, ...groups.sites]);
    const commands = program.commands
        .filter(command => !adapterNames.has(command.name()))
        .map(command => ({
        name: command.name(),
        description: command.description(),
    }));
    const sortLocale = (a, b) => a.localeCompare(b);
    return {
        name: program.name(),
        description: program.description(),
        commands,
        external_clis: {
            count: groups.external.length,
            clis: groups.external.map(cli => cli.name).sort(sortLocale),
            display: groups.external.map(cli => cli.label).sort(sortLocale),
        },
        app_adapters: {
            count: groups.apps.length,
            apps: [...groups.apps].sort(sortLocale),
        },
        site_adapters: {
            count: groups.sites.length,
            sites: [...groups.sites].sort(sortLocale),
        },
        next: [
            `${CLI_COMMAND} <site> --help -f yaml`,
            `${CLI_COMMAND} list -f yaml`,
            `${CLI_COMMAND} <site> <command> -f yaml`,
        ],
    };
}
export function siteHelpData(site, commands) {
    return buildSiteHelpData(site, commands.map(toPresentableCommand));
}
export function commandHelpData(cmd) {
    return buildCommandHelpData(toPresentableCommand(cmd));
}
export function formatCommonOptionsHelpText() {
    return formatCommonOptionsHelp();
}
export function formatBrowserCommonOptionsHelpText() {
    return formatBrowserCommonOptionsHelp();
}
export function formatSiteHelpText(site, commands) {
    return formatSiteHelp(site, commands.map(toPresentableCommand));
}
export function formatCommandHelpText(cmd) {
    return formatCommandHelp(toPresentableCommand(cmd));
}
export function installStructuredHelp(command, data, textSuffix) {
    const original = command.helpInformation.bind(command);
    command.helpInformation = ((contextOptions) => {
        const format = getRequestedHelpFormat();
        if (format)
            return renderStructuredHelp(data(), format);
        const suffix = typeof textSuffix === 'function' ? textSuffix() : textSuffix ?? '';
        return original(contextOptions) + suffix;
    });
}
export function formatSiteCommandDescription(cmd) {
    return buildSiteCommandDescription(toPresentableCommand(cmd));
}
