// @agentrhq/webcmd/dist/src/session-lease.js
import { AsyncLocalStorage } from 'node:async_hooks';
/** Inactivity window after which an unowned session lease expires. */
export const SESSION_LEASE_TTL_MS = 45_000;
let runIdCounter = 0;
/** Generate an id for one complete logical CLI command run. */
export function generateRunId() {
    return `run_${process.pid}_${Date.now()}_${++runIdCounter}`;
}
let activeRun;
const daemonRunContextStorage = new AsyncLocalStorage();
/** Run one logical execution with context isolated across its async chain. */
export function runWithDaemonRunContext(context, callback) {
    return daemonRunContextStorage.run(context, callback);
}
export function setDaemonRunContext(context) {
    activeRun = context;
}
export function getDaemonRunContext() {
    return daemonRunContextStorage.getStore() ?? activeRun;
}
/**
 * Clear only the context still owned by `runId`. Deferred cleanup from an old
 * command must not clear a newer command's run identity.
 */
export function clearDaemonRunContext(runId) {
    if (activeRun?.runId === runId)
        activeRun = undefined;
}
const UNKNOWN_OUTCOME_CODES = [
    'command_result_unknown',
    'command_lost',
    'result_evicted',
];
function containsUnknownOutcomeCode(value) {
    if (typeof value !== 'string')
        return false;
    const normalized = value.toLowerCase();
    return UNKNOWN_OUTCOME_CODES.some((code) => {
        const start = normalized.indexOf(code);
        if (start < 0)
            return false;
        const before = normalized[start - 1];
        const after = normalized[start + code.length];
        return (!before || !/[a-z0-9_]/.test(before)) && (!after || !/[a-z0-9_]/.test(after));
    });
}
/**
 * Detect errors whose browser-side result is unknown. The traversal includes
 * wrapper causes and AggregateError members and tolerates cyclic error graphs.
 */
export function isUnknownOutcomeError(error) {
    const queue = [error];
    const seen = new Set();
    while (queue.length > 0) {
        const current = queue.pop();
        if (!current || (typeof current !== 'object' && typeof current !== 'function') || seen.has(current)) {
            continue;
        }
        seen.add(current);
        const record = current;
        if (containsUnknownOutcomeCode(record.code)
            || containsUnknownOutcomeCode(record.errorCode)
            || containsUnknownOutcomeCode(record.daemonCode)
            || containsUnknownOutcomeCode(record.message)
            || containsUnknownOutcomeCode(record.error)) {
            return true;
        }
        if (record.cause !== undefined)
            queue.push(record.cause);
        if (Array.isArray(record.errors))
            queue.push(...record.errors);
    }
    return false;
}
/**
 * Lease key for a site session after the daemon has resolved its actual Cloak
 * profile. Encoding the session keeps key partitions unambiguous.
 */
export function getSessionLeaseKey(profileId, surface, session) {
    return `${profileId}␟${surface}␟${encodeURIComponent(session)}`;
}
/** Whether a process id is safe to interpolate into local process guidance. */
export function isActionablePid(pid) {
    return typeof pid === 'number' && Number.isSafeInteger(pid) && pid > 0;
}
function pidFromRunId(runId) {
    const match = /^run_(\d+)_/.exec(runId);
    if (!match)
        return undefined;
    const pid = Number(match[1]);
    return isActionablePid(pid) ? pid : undefined;
}
/** Only persistent adapter writes with a complete owner identity need a lease. */
export function isSessionLeaseCommand(command) {
    return command.surface === 'adapter'
        && command.siteSession === 'persistent'
        && command.access === 'write'
        && typeof command.session === 'string'
        && command.session.length > 0
        && typeof command.runId === 'string'
        && command.runId.length > 0;
}
export class SessionLeaseRegistry {
    now;
    leases = new Map();
    constructor(now = Date.now) {
        this.now = now;
    }
    acquire(input, hasPendingWork) {
        const now = this.now();
        const current = this.leases.get(input.key);
        const currentIsLive = current !== undefined
            && (now - current.heartbeatAt <= SESSION_LEASE_TTL_MS || hasPendingWork(current.runId));
        if (current && currentIsLive && current.runId !== input.runId) {
            return { acquired: false, holder: { ...current } };
        }
        const pid = input.pid === undefined
            ? pidFromRunId(input.runId)
            : (isActionablePid(input.pid) ? input.pid : undefined);
        const lease = current?.runId === input.runId
            ? { ...current, heartbeatAt: now }
            : {
                key: input.key,
                runId: input.runId,
                command: input.command,
                ...(pid === undefined ? {} : { pid }),
                acquiredAt: now,
                heartbeatAt: now,
            };
        this.leases.set(input.key, lease);
        return { acquired: true, lease: { ...lease } };
    }
    /** Refresh a lease only when both its key and logical owner match. */
    heartbeat(key, runId) {
        const current = this.leases.get(key);
        if (!current || current.runId !== runId)
            return false;
        current.heartbeatAt = this.now();
        return true;
    }
    /** Release all leases owned by one logical run. */
    releaseByRunId(runId) {
        let released = 0;
        for (const [key, lease] of this.leases) {
            if (lease.runId !== runId)
                continue;
            this.leases.delete(key);
            released += 1;
        }
        return released;
    }
    /** Return a snapshot of currently live leases without exposing mutable state. */
    list(hasPendingWork) {
        const now = this.now();
        const active = [];
        for (const lease of this.leases.values()) {
            if (now - lease.heartbeatAt <= SESSION_LEASE_TTL_MS || hasPendingWork(lease.runId)) {
                active.push({ ...lease });
            }
        }
        return active;
    }
}
