// @agentrhq/webcmd/dist/src/browser/runtime/local-cloak/session-manager.js
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { execFile } from 'node:child_process';
import { launchPersistentContext as cloakLaunchPersistentContext } from 'cloakbrowser';
import { activateDarwinBackgroundContext, launchDarwinBackgroundPersistentContext } from './darwin-background-launch.js';
import { normalizeProfileId, resolveCloakProfileDir } from './profiles.js';
import { CloakNetworkCapture } from './network.js';
import { findPackageRoot } from '../../../package-paths.js';
/** Installed `cloakbrowser` npm package version, for doctor/status display. */
export function resolveCloakBrowserVersion() {
    try {
        const entryPath = fileURLToPath(import.meta.resolve('cloakbrowser'));
        const pkg = JSON.parse(fs.readFileSync(path.join(findPackageRoot(entryPath), 'package.json'), 'utf-8'));
        return typeof pkg.version === 'string' ? pkg.version : undefined;
    }
    catch {
        return undefined;
    }
}
let pageCounter = 0;
export function resolveLeaseKey(input) {
    const surface = input.surface === 'adapter' ? 'adapter' : 'browser';
    const session = input.session?.trim();
    if (!session)
        throw new Error('Browser session is required.');
    return `${surface}\u0000${encodeURIComponent(session)}`;
}
function pageIsClosed(page) {
    return page.isClosed?.() === true;
}
function isClosedContextError(error) {
    const message = error instanceof Error ? error.message : String(error);
    return /Target page, context or browser has been closed/i.test(message);
}
export class CloakSessionManager {
    opts;
    networkCapture = new CloakNetworkCapture();
    launchPersistentContext;
    launchBackgroundPersistentContext;
    activateBackgroundContext;
    platform;
    recoverLockedProfile;
    profiles = new Map();
    profileLaunches = new Map();
    constructor(opts = {}) {
        this.opts = opts;
        this.launchPersistentContext = opts.launchPersistentContext ?? cloakLaunchPersistentContext;
        this.launchBackgroundPersistentContext = opts.launchBackgroundPersistentContext ?? launchDarwinBackgroundPersistentContext;
        this.activateBackgroundContext = opts.activateBackgroundContext ?? activateDarwinBackgroundContext;
        this.platform = opts.platform ?? process.platform;
        this.recoverLockedProfile = opts.recoverLockedProfile ?? recoverLockedCloakProfile;
    }
    profileStatuses() {
        return [...this.profiles.entries()].map(([contextId, runtime]) => ({
            contextId,
            runtimeConnected: true,
            runtimeVersion: resolveCloakBrowserVersion(),
            pending: 0,
            lastSeenAt: runtime.lastSeenAt,
        }));
    }
    activeProfileIds() {
        return [...this.profiles.keys()];
    }
    async getPage(input) {
        const profileId = normalizeProfileId(input.profileId);
        const session = requireSession(input.session);
        const surface = normalizeSurface(input.surface);
        const leaseKey = resolveLeaseKey(input);
        const runtime = await this.getProfileRuntime(profileId, input.windowMode);
        const freshPage = input.freshPage === true;
        const existing = runtime.pages.get(leaseKey);
        if (existing && !pageIsClosed(existing.page) && !freshPage) {
            runtime.lastSeenAt = Date.now();
            existing.idleTimeout = input.idleTimeout;
            this.refreshIdleTimer(runtime, leaseKey, existing);
            return { profileId, leaseKey, context: runtime.context, page: existing.page, pageId: existing.pageId };
        }
        if (existing && freshPage) {
            runtime.pages.delete(leaseKey);
            this.clearIdleTimer(existing);
            if (runtime.selectedPageId === existing.pageId)
                runtime.selectedPageId = undefined;
            if (!pageIsClosed(existing.page))
                await existing.page.close().catch(() => { });
        }
        return this.createPageWithRecovery(profileId, input.windowMode, (candidate) => {
            const existingPages = candidate.context.pages();
            // freshPage must never adopt a leftover tab — its whole point is a clean DOM.
            return !freshPage && existingPages[0] && candidate.pages.size === 0
                ? existingPages[0]
                : candidate.context.newPage();
        }, (candidate, page) => {
            const pageId = nextPageId();
            const entry = { page, pageId, session, surface, siteSession: input.siteSession, idleTimeout: input.idleTimeout };
            candidate.pages.set(leaseKey, entry);
            this.refreshIdleTimer(candidate, leaseKey, entry);
            candidate.selectedPageId = pageId;
            candidate.lastSeenAt = Date.now();
            return { profileId, leaseKey, context: candidate.context, page, pageId };
        });
    }
    findPageById(pageId, opts = {}) {
        for (const [profileId, runtime] of this.profiles.entries()) {
            for (const [leaseKey, entry] of runtime.pages.entries()) {
                if (entry.pageId === pageId && !pageIsClosed(entry.page)) {
                    entry.idleTimeout = opts.idleTimeout;
                    this.refreshIdleTimer(runtime, leaseKey, entry);
                    return { profileId, leaseKey, context: runtime.context, page: entry.page, pageId: entry.pageId };
                }
            }
        }
        return null;
    }
    pageIdFor(page) {
        for (const runtime of this.profiles.values()) {
            for (const entry of runtime.pages.values()) {
                if (entry.page === page)
                    return entry.pageId;
            }
        }
        return undefined;
    }
    registerPage(input, page) {
        const profileId = normalizeProfileId(input.profileId);
        const session = requireSession(input.session);
        const surface = normalizeSurface(input.surface);
        const runtime = this.profiles.get(profileId);
        if (!runtime)
            throw new Error(`Profile ${profileId} is not running`);
        const pageId = nextPageId();
        const leaseKey = `${resolveLeaseKey(input)}\u0000${pageId}`;
        const entry = { page, pageId, session, surface, siteSession: input.siteSession, idleTimeout: input.idleTimeout };
        runtime.pages.set(leaseKey, entry);
        this.refreshIdleTimer(runtime, leaseKey, entry);
        runtime.selectedPageId = pageId;
        runtime.lastSeenAt = Date.now();
        return pageId;
    }
    async listPages(input) {
        const profileId = normalizeProfileId(input.profileId);
        const runtime = this.profiles.get(profileId);
        if (!runtime)
            return [];
        const entries = this.openEntries(runtime);
        return Promise.all(entries.map(async ([, entry], index) => ({
            id: entry.pageId,
            page: entry.pageId,
            index,
            title: await entry.page.title().catch(() => ''),
            url: entry.page.url(),
            profileId,
            session: entry.session,
            surface: entry.surface,
            selected: runtime.selectedPageId === entry.pageId,
        })));
    }
    async newPage(input) {
        const profileId = normalizeProfileId(input.profileId);
        const session = requireSession(input.session);
        const surface = normalizeSurface(input.surface);
        const acquired = await this.createPageWithRecovery(profileId, input.windowMode, (candidate) => candidate.context.newPage(), (runtime, page) => ({ runtime, page }));
        if (input.url) {
            try {
                await acquired.page.goto(input.url, { waitUntil: 'load' });
            }
            catch (error) {
                if (!pageIsClosed(acquired.page))
                    await acquired.page.close().catch(() => { });
                throw error;
            }
        }
        if (this.profiles.get(profileId) !== acquired.runtime) {
            if (!pageIsClosed(acquired.page))
                await acquired.page.close().catch(() => { });
            throw new Error('Target page, context or browser has been closed');
        }
        const pageId = nextPageId();
        const leaseKey = `${resolveLeaseKey(input)}\u0000${pageId}`;
        const entry = {
            page: acquired.page,
            pageId,
            session,
            surface,
            siteSession: input.siteSession,
            idleTimeout: input.idleTimeout,
        };
        acquired.runtime.pages.set(leaseKey, entry);
        this.refreshIdleTimer(acquired.runtime, leaseKey, entry);
        acquired.runtime.lastSeenAt = Date.now();
        return { profileId, leaseKey, context: acquired.runtime.context, page: acquired.page, pageId };
    }
    async selectPage(input) {
        const profileId = normalizeProfileId(input.profileId);
        const runtime = this.profiles.get(profileId);
        if (!runtime)
            return null;
        const match = input.pageId ? this.findEntryByPageId(runtime, input.pageId) : this.openEntries(runtime)[input.index ?? -1];
        if (!match)
            return null;
        const [leaseKey, entry] = match;
        if (input.windowMode !== 'background') {
            await entry.page.bringToFront?.().catch(() => { });
            await this.activateBackgroundContext(runtime.context);
        }
        runtime.selectedPageId = entry.pageId;
        runtime.lastSeenAt = Date.now();
        return { profileId, leaseKey, context: runtime.context, page: entry.page, pageId: entry.pageId };
    }
    async bindPage(input) {
        const profileId = normalizeProfileId(input.profileId);
        const session = requireSession(input.session);
        const surface = normalizeSurface(input.surface);
        const runtime = this.profiles.get(profileId);
        if (!runtime)
            return null;
        const match = input.pageId ? this.findEntryByPageId(runtime, input.pageId) : this.openEntries(runtime)[input.index ?? -1];
        if (!match)
            return null;
        const [sourceKey, entry] = match;
        const canonicalKey = resolveLeaseKey({ profileId, session, surface });
        const currentCanonical = runtime.pages.get(canonicalKey);
        if (currentCanonical && currentCanonical !== entry && !pageIsClosed(currentCanonical.page)) {
            const preservedKey = `${canonicalKey}\u0000${currentCanonical.pageId}`;
            runtime.pages.delete(canonicalKey);
            runtime.pages.set(preservedKey, currentCanonical);
            this.refreshIdleTimer(runtime, preservedKey, currentCanonical);
        }
        if (sourceKey !== canonicalKey)
            runtime.pages.delete(sourceKey);
        entry.session = session;
        entry.surface = surface;
        entry.siteSession = input.siteSession;
        entry.idleTimeout = input.idleTimeout;
        runtime.pages.set(canonicalKey, entry);
        if (input.windowMode !== 'background') {
            await entry.page.bringToFront?.().catch(() => { });
            await this.activateBackgroundContext(runtime.context);
        }
        this.refreshIdleTimer(runtime, canonicalKey, entry);
        runtime.selectedPageId = entry.pageId;
        runtime.lastSeenAt = Date.now();
        return { profileId, leaseKey: canonicalKey, context: runtime.context, page: entry.page, pageId: entry.pageId };
    }
    async closePage(input) {
        const profileId = normalizeProfileId(input.profileId);
        const runtime = this.profiles.get(profileId);
        if (!runtime)
            return null;
        const match = input.pageId ? this.findEntryByPageId(runtime, input.pageId) : this.openEntries(runtime)[input.index ?? -1];
        if (!match)
            return null;
        const [leaseKey, entry] = match;
        runtime.pages.delete(leaseKey);
        this.clearIdleTimer(entry);
        if (runtime.selectedPageId === entry.pageId)
            runtime.selectedPageId = undefined;
        if (!pageIsClosed(entry.page))
            await entry.page.close().catch(() => { });
        runtime.lastSeenAt = Date.now();
        return entry.pageId;
    }
    async release(input) {
        const profileId = normalizeProfileId(input.profileId);
        const runtime = this.profiles.get(profileId);
        if (!runtime)
            return;
        const leaseKey = resolveLeaseKey(input);
        const exactEntry = runtime.pages.get(leaseKey);
        const entries = exactEntry
            ? [[leaseKey, exactEntry]]
            : this.openEntries(runtime).filter(([, entry]) => entry.session === requireSession(input.session) && entry.surface === normalizeSurface(input.surface));
        for (const [key, entry] of entries) {
            runtime.pages.delete(key);
            this.clearIdleTimer(entry);
            if (runtime.selectedPageId === entry.pageId)
                runtime.selectedPageId = undefined;
            if (entry.siteSession !== 'persistent' && !pageIsClosed(entry.page)) {
                await entry.page.close().catch(() => { });
            }
        }
    }
    async shutdown() {
        for (const runtime of this.profiles.values()) {
            for (const entry of runtime.pages.values())
                this.clearIdleTimer(entry);
            await runtime.context.close().catch(() => { });
        }
        this.profiles.clear();
    }
    async getProfileRuntime(profileId, windowMode) {
        const existing = this.profiles.get(profileId);
        if (existing)
            return existing;
        const pending = this.profileLaunches.get(profileId);
        if (pending)
            return pending;
        const launch = this.launchProfileRuntime(profileId, windowMode);
        this.profileLaunches.set(profileId, launch);
        try {
            return await launch;
        }
        finally {
            this.profileLaunches.delete(profileId);
        }
    }
    async launchProfileRuntime(profileId, windowMode) {
        const userDataDir = resolveCloakProfileDir(profileId, { baseDir: this.opts.baseDir });
        fs.mkdirSync(userDataDir, { recursive: true });
        const launchOptions = {
            userDataDir,
            headless: false,
            humanize: true,
        };
        const launchPersistentContext = this.platform === 'darwin' && windowMode === 'background'
            ? this.launchBackgroundPersistentContext
            : this.launchPersistentContext;
        let context;
        try {
            context = await launchPersistentContext(launchOptions);
        }
        catch (err) {
            if (!isProfileAlreadyInUseError(err) || !(await this.recoverLockedProfile(userDataDir)))
                throw err;
            context = await launchPersistentContext(launchOptions);
        }
        const runtime = { context, pages: new Map(), lastSeenAt: Date.now() };
        this.attachRuntimeLifecycle(profileId, runtime);
        this.profiles.set(profileId, runtime);
        return runtime;
    }
    invalidateProfileRuntime(profileId, runtime) {
        if (this.profiles.get(profileId) !== runtime)
            return;
        this.profiles.delete(profileId);
        for (const entry of runtime.pages.values()) {
            if (entry.idleTimer)
                clearTimeout(entry.idleTimer);
            this.networkCapture.stop(entry.page);
        }
        runtime.pages.clear();
        runtime.selectedPageId = undefined;
    }
    attachRuntimeLifecycle(profileId, runtime) {
        runtime.context.on('close', () => this.invalidateProfileRuntime(profileId, runtime));
    }
    async createPageWithRecovery(profileId, windowMode, createPage, commitPage, attempt = 0) {
        const runtime = await this.getProfileRuntime(profileId, windowMode);
        let page;
        try {
            page = await createPage(runtime);
        }
        catch (error) {
            if (attempt !== 0 || !isClosedContextError(error))
                throw error;
            this.invalidateProfileRuntime(profileId, runtime);
            return this.createPageWithRecovery(profileId, windowMode, createPage, commitPage, 1);
        }
        if (this.profiles.get(profileId) !== runtime) {
            if (!pageIsClosed(page))
                await page.close().catch(() => { });
            throw new Error('Target page, context or browser has been closed');
        }
        return commitPage(runtime, page);
    }
    openEntries(runtime) {
        return [...runtime.pages.entries()].filter(([, entry]) => !pageIsClosed(entry.page));
    }
    findEntryByPageId(runtime, pageId) {
        return this.openEntries(runtime).find(([, entry]) => entry.pageId === pageId) ?? null;
    }
    refreshIdleTimer(runtime, leaseKey, entry) {
        this.clearIdleTimer(entry);
        if (!entry.idleTimeout || entry.idleTimeout <= 0 || entry.siteSession === 'persistent')
            return;
        entry.idleTimer = setTimeout(() => {
            void this.expireLease(runtime, leaseKey, entry);
        }, entry.idleTimeout);
        entry.idleTimer.unref?.();
    }
    async expireLease(runtime, leaseKey, entry) {
        if (runtime.pages.get(leaseKey) !== entry)
            return;
        runtime.pages.delete(leaseKey);
        this.clearIdleTimer(entry);
        if (runtime.selectedPageId === entry.pageId)
            runtime.selectedPageId = undefined;
        runtime.lastSeenAt = Date.now();
        if (entry.siteSession !== 'persistent' && !pageIsClosed(entry.page)) {
            await entry.page.close().catch(() => { });
        }
    }
    clearIdleTimer(entry) {
        if (entry.idleTimer)
            clearTimeout(entry.idleTimer);
        entry.idleTimer = undefined;
    }
}
function nextPageId() {
    return `page-${Date.now()}-${++pageCounter}`;
}
function normalizeSurface(surface) {
    return surface === 'adapter' ? 'adapter' : 'browser';
}
function requireSession(session) {
    const normalized = session?.trim();
    if (!normalized)
        throw new Error('Browser session is required.');
    return normalized;
}
function isProfileAlreadyInUseError(err) {
    const message = err instanceof Error ? err.message : String(err);
    return message.includes('Opening in existing browser session')
        || message.includes('Failed to create a ProcessSingleton for your profile directory');
}
async function recoverLockedCloakProfile(userDataDir) {
    if (process.platform === 'win32')
        return false;
    const initial = await findCloakProfileProcesses(userDataDir);
    if (initial.length === 0)
        return false;
    signalPids(initial, 'SIGTERM');
    if (await waitForProfileProcessesToExit(userDataDir, 2500))
        return true;
    signalPids(await findCloakProfileProcesses(userDataDir), 'SIGKILL');
    return waitForProfileProcessesToExit(userDataDir, 1500);
}
async function waitForProfileProcessesToExit(userDataDir, timeoutMs) {
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
        await new Promise((resolve) => setTimeout(resolve, 100));
        if ((await findCloakProfileProcesses(userDataDir)).length === 0)
            return true;
    }
    return (await findCloakProfileProcesses(userDataDir)).length === 0;
}
function signalPids(pids, signal) {
    for (const pid of pids) {
        try {
            process.kill(pid, signal);
        }
        catch {
            // Already exited or not signalable; the follow-up poll decides recovery.
        }
    }
}
async function findCloakProfileProcesses(userDataDir) {
    const profileDirs = profileDirAliases(userDataDir);
    const stdout = await psOutput();
    const pids = [];
    for (const line of stdout.split('\n')) {
        const match = line.match(/^\s*(\d+)\s+(.+)$/);
        if (!match)
            continue;
        const pid = Number(match[1]);
        const command = match[2];
        if (!Number.isInteger(pid) || pid === process.pid)
            continue;
        if (!isCloakBrowserCommand(command))
            continue;
        if (!commandUsesProfileDir(command, profileDirs))
            continue;
        pids.push(pid);
    }
    return [...new Set(pids)];
}
function commandUsesProfileDir(command, profileDirs) {
    for (const dir of profileDirs) {
        const marker = `--user-data-dir=${dir}`;
        const index = command.indexOf(marker);
        if (index < 0)
            continue;
        const next = command[index + marker.length];
        if (next === undefined || /\s/.test(next))
            return true;
    }
    return false;
}
function profileDirAliases(userDataDir) {
    const aliases = new Set([userDataDir]);
    try {
        aliases.add(fs.realpathSync.native(userDataDir));
    }
    catch {
        // The launch path is still useful even if realpath cannot resolve it.
    }
    return [...aliases];
}
function isCloakBrowserCommand(command) {
    return command.includes('/.cloakbrowser/') || command.includes('\\.cloakbrowser\\');
}
function psOutput() {
    return new Promise((resolve) => {
        execFile('ps', ['-axo', 'pid=,command='], { encoding: 'utf8', maxBuffer: 10 * 1024 * 1024, timeout: 2000 }, (err, stdout) => {
            resolve(err ? '' : String(stdout));
        });
    });
}
