// @agentrhq/webcmd/dist/src/hosted/config.js
import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import { CONFIG_DIR_NAME, ENV_PREFIX } from '../brand.js';
export function getConfigDir(io = {}) {
    const env = io.env ?? process.env;
    return env[`${ENV_PREFIX}_CONFIG_DIR`] || path.join(io.homeDir ?? os.homedir(), CONFIG_DIR_NAME);
}
export function getConfigPath(io = {}) {
    return path.join(getConfigDir(io), 'config.json');
}
function parseConfig(raw) {
    const parsed = JSON.parse(raw);
    if (parsed.mode === 'local' && typeof parsed.updatedAt === 'string') {
        return { mode: 'local', updatedAt: parsed.updatedAt };
    }
    if (parsed.mode === 'hosted'
        && typeof parsed.updatedAt === 'string'
        && typeof parsed.hosted?.apiBaseUrl === 'string'
        && (typeof parsed.hosted?.apiKey === 'string' || typeof parsed.hosted?.apiKeyRef === 'string')) {
        const credentialBackend = readCredentialBackend(parsed.hosted.credentialBackend);
        return {
            mode: 'hosted',
            updatedAt: parsed.updatedAt,
            hosted: {
                apiBaseUrl: parsed.hosted.apiBaseUrl,
                ...(typeof parsed.hosted.apiKey === 'string' ? { apiKey: parsed.hosted.apiKey } : {}),
                ...(typeof parsed.hosted.apiKeyRef === 'string' ? { apiKeyRef: parsed.hosted.apiKeyRef } : {}),
                ...(credentialBackend ? { credentialBackend } : {}),
                ...(parsed.hosted.manifestCache ? { manifestCache: parsed.hosted.manifestCache } : {}),
            },
        };
    }
    return { mode: 'local', updatedAt: new Date(0).toISOString() };
}
export function loadWebcmdConfig(io = {}) {
    const readFileSync = io.readFileSync ?? fs.readFileSync;
    try {
        return parseConfig(readFileSync(getConfigPath(io), 'utf-8'));
    }
    catch {
        return { mode: 'local', updatedAt: new Date(0).toISOString() };
    }
}
export function saveWebcmdConfig(config, io = {}) {
    const writeFileSync = io.writeFileSync ?? fs.writeFileSync;
    const mkdirSync = io.mkdirSync ?? fs.mkdirSync;
    const chmodSync = io.chmodSync ?? fs.chmodSync;
    const target = getConfigPath(io);
    mkdirSync(path.dirname(target), { recursive: true });
    writeFileSync(target, `${JSON.stringify(persistableConfig(config), null, 2)}\n`, { encoding: 'utf-8', mode: 0o600 });
    try {
        chmodSync(target, 0o600);
    }
    catch {
        // Windows and unusual filesystems may not support POSIX modes.
    }
}
export function makeLocalConfig(now = new Date()) {
    return {
        mode: 'local',
        updatedAt: now.toISOString(),
    };
}
export function makeHostedConfig(input) {
    return {
        mode: 'hosted',
        updatedAt: (input.now ?? new Date()).toISOString(),
        hosted: {
            apiBaseUrl: normalizeApiBaseUrl(input.apiBaseUrl),
            ...(input.apiKey !== undefined ? { apiKey: input.apiKey.trim() } : {}),
            ...(input.apiKeyRef !== undefined ? { apiKeyRef: input.apiKeyRef } : {}),
            ...(input.credentialBackend !== undefined ? { credentialBackend: input.credentialBackend } : {}),
            ...(input.manifestCache ? { manifestCache: input.manifestCache } : {}),
        },
    };
}
export function normalizeApiBaseUrl(raw) {
    const value = raw.trim().replace(/\/+$/, '');
    return value || defaultHostedApiBaseUrl();
}
export function defaultHostedApiBaseUrl(env = process.env) {
    return normalizeConfiguredUrl(env.WEBCMD_CLOUD_API_URL) ?? 'https://api.webcmd.dev';
}
function normalizeConfiguredUrl(raw) {
    if (!raw?.trim())
        return undefined;
    return raw.trim().replace(/\/+$/, '');
}
export function isHostedConfig(config) {
    return config.mode === 'hosted';
}
export function shouldUseHostedMode(io = {}) {
    return isHostedConfig(loadWebcmdConfig(io));
}
function readCredentialBackend(value) {
    return value === 'os' || value === 'file-fallback' ? value : undefined;
}
function persistableConfig(config) {
    if (config.mode !== 'hosted' || !config.hosted.apiKeyRef)
        return config;
    return {
        mode: 'hosted',
        updatedAt: config.updatedAt,
        hosted: {
            apiBaseUrl: config.hosted.apiBaseUrl,
            apiKeyRef: config.hosted.apiKeyRef,
            ...(config.hosted.credentialBackend ? { credentialBackend: config.hosted.credentialBackend } : {}),
            ...(config.hosted.manifestCache ? { manifestCache: config.hosted.manifestCache } : {}),
        },
    };
}
