// @agentrhq/webcmd/dist/src/hosted/manifest.js
import { commandHelpData, commandListPresentation, commandListRows, formatCommandHelp, formatSiteHelp, getCommandCompletionCandidates, siteHelpData, toPresentableCommand, } from '../command-presentation.js';
export function isLocalOnlyHostedCommand(command) {
    return command.strategy.toUpperCase() === 'LOCAL';
}
export function hostedCommands(manifest) {
    return manifest.commands
        .filter((command) => !isLocalOnlyHostedCommand(command))
        .sort((a, b) => a.command.localeCompare(b.command));
}
export function findHostedCommand(manifest, site, name) {
    return manifest.commands.find((command) => {
        return command.site === site && (command.name === name || command.aliases?.includes(name));
    }) ?? null;
}
export function presentHostedCommand(command) {
    return toPresentableCommand(command);
}
export function hostedListRows(manifest, structured) {
    return commandListRows(hostedCommands(manifest).map(presentHostedCommand), structured);
}
export function hostedListPresentation(manifest, format) {
    return commandListPresentation(hostedCommands(manifest).map(presentHostedCommand), format);
}
export function siteNames(manifest) {
    return getCommandCompletionCandidates(hostedCommands(manifest), [], 1, []);
}
export function commandNamesForSite(manifest, site) {
    return getCommandCompletionCandidates(hostedCommands(manifest), [site], 2, []);
}
export function renderHostedSiteHelp(manifest, site) {
    const commands = hostedCommands(manifest).filter((command) => command.site === site);
    if (commands.length === 0)
        return `Unknown hosted Webcmd site: ${site}\n`;
    return formatSiteHelp(site, commands.map(presentHostedCommand));
}
export function hostedSiteHelpData(manifest, site) {
    const commands = hostedCommands(manifest).filter((command) => command.site === site);
    if (commands.length === 0)
        return null;
    return siteHelpData(site, commands.map(presentHostedCommand));
}
export function renderHostedCommandHelp(command) {
    return formatCommandHelp(presentHostedCommand(command));
}
export function hostedCommandHelpData(command) {
    return commandHelpData(presentHostedCommand(command));
}
