// @agentrhq/webcmd/dist/src/browser/daemon-transport.js
import { DAEMON_HEADER_NAME, DEFAULT_DAEMON_PORT } from '../constants.js';
const DAEMON_PORT = DEFAULT_DAEMON_PORT;
const DAEMON_URL = `http://127.0.0.1:${DAEMON_PORT}`;
const WEBCMD_HEADERS = { [DAEMON_HEADER_NAME]: '1' };
export async function requestDaemon(pathname, init) {
    const { timeout = 2000, headers, ...rest } = init ?? {};
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeout);
    try {
        return await fetch(`${DAEMON_URL}${pathname}`, {
            ...rest,
            headers: { ...WEBCMD_HEADERS, ...headers },
            signal: controller.signal,
        });
    }
    finally {
        clearTimeout(timer);
    }
}
export async function fetchDaemonStatus(opts) {
    try {
        const params = opts?.contextId ? `?contextId=${encodeURIComponent(opts.contextId)}` : '';
        const res = await requestDaemon(`/status${params}`, { timeout: opts?.timeout ?? 2000 });
        if (!res.ok)
            return null;
        return await res.json();
    }
    catch {
        return null;
    }
}
export async function getDaemonHealth(opts) {
    const status = await fetchDaemonStatus(opts);
    if (!status)
        return { state: 'stopped', status: null };
    if (status.profileRequired)
        return { state: 'profile-required', status };
    if (status.profileDisconnected)
        return { state: 'profile-disconnected', status };
    if (!status.runtimeConnected)
        return { state: 'no-runtime', status };
    return { state: 'ready', status };
}
export async function requestDaemonShutdown(opts) {
    try {
        const res = await requestDaemon('/shutdown', { method: 'POST', timeout: opts?.timeout ?? 5000 });
        return res.ok;
    }
    catch {
        return false;
    }
}
