// @agentrhq/webcmd/dist/src/browser/daemon-client.js
/**
 * HTTP client for communicating with the webcmd daemon.
 *
 * Provides a typed send() function that posts a Command and returns a Result.
 */
import { sleep } from '../utils.js';
import { BrowserConnectError, SessionBusyError } from '../errors.js';
import { COMMAND_RESULT_UNKNOWN_CODE, COMMAND_RESULT_UNKNOWN_HINT } from '../daemon-utils.js';
import { getDaemonRunContext } from '../session-lease.js';
import { classifyBrowserError } from './errors.js';
import { profileRouteParams, resolveProfileSelection } from './profile.js';
import { DEFAULT_BROWSER_CONNECT_TIMEOUT } from './config.js';
import { ensureBrowserBridgeReady } from './daemon-lifecycle.js';
import { isPreDispatchError } from './bridge-readiness.js';
import { fetchDaemonStatus, getDaemonHealth, requestDaemon, requestDaemonShutdown, } from './daemon-transport.js';
let _idCounter = 0;
function generateId() {
    return `cmd_${process.pid}_${Date.now()}_${++_idCounter}`;
}
const DEFAULT_COMMAND_TIMEOUT_SECONDS = 120;
const RUNTIME_OP_TIMEOUT_MARGIN_MS = 15_000;
const HTTP_TIMEOUT_MARGIN_MS = 10_000;
const TRANSPORT_MAX_ATTEMPTS = 4;
let _userCommandTimeoutSeconds = null;
export function setDaemonCommandTimeoutSeconds(seconds) {
    _userCommandTimeoutSeconds = typeof seconds === 'number' && seconds > 0 ? Math.ceil(seconds) : null;
}
function effectiveCommandTimeoutSeconds(params) {
    const base = _userCommandTimeoutSeconds ?? DEFAULT_COMMAND_TIMEOUT_SECONDS;
    if (typeof params.timeoutMs === 'number' && params.timeoutMs > 0) {
        return Math.max(base, Math.ceil((params.timeoutMs + RUNTIME_OP_TIMEOUT_MARGIN_MS) / 1000));
    }
    return base;
}
const UNKNOWN_OUTCOME_CODES = new Set([
    COMMAND_RESULT_UNKNOWN_CODE,
    'command_lost',
    'result_evicted',
]);
const PRE_CONNECT_ERROR_CODES = new Set([
    'ECONNREFUSED',
    'UND_ERR_CONNECT_TIMEOUT',
    'EHOSTUNREACH',
    'ENETUNREACH',
    'ENOTFOUND',
]);
function isPreConnectFetchError(err) {
    const queue = [err];
    const seen = new Set();
    while (queue.length) {
        const current = queue.pop();
        if (!current || typeof current !== 'object' || seen.has(current))
            continue;
        seen.add(current);
        const { code, cause, errors } = current;
        if (typeof code === 'string' && PRE_CONNECT_ERROR_CODES.has(code))
            return true;
        if (cause)
            queue.push(cause);
        if (Array.isArray(errors))
            queue.push(...errors);
    }
    return false;
}
export class BrowserCommandError extends Error {
    code;
    hint;
    constructor(message, code, hint) {
        super(message);
        this.code = code;
        this.hint = hint;
        this.name = 'BrowserCommandError';
    }
}
export { fetchDaemonStatus, getDaemonHealth, requestDaemonShutdown, };
/**
 * Internal: send a command to the daemon and return the raw `DaemonResult`.
 *
 * Retry policy is explicit:
 * - pre-dispatch bridge/profile errors and pre-connect fetch failures run the
 *   full daemon/runtime ensure path, then resend the same command id;
 * - executor-transient errors that happened before page code ran get one new
 *   logical attempt with a fresh id;
 * - `command_result_unknown`, duplicate pending ids from old daemons,
 *   post-connect drops, and AbortError are never retried automatically.
 */
async function sendCommandRaw(action, params) {
    const timeoutSeconds = effectiveCommandTimeoutSeconds(params);
    const deadlineAt = Date.now() + timeoutSeconds * 1000;
    const rawWindowMode = process.env.WEBCMD_WINDOW;
    const envWindowMode = rawWindowMode === 'foreground' || rawWindowMode === 'background'
        ? rawWindowMode
        : undefined;
    const routing = params.contextId || params.preferredContextId
        ? { contextId: params.contextId, preferredContextId: params.preferredContextId }
        : profileRouteParams(resolveProfileSelection());
    const contextId = routing.contextId;
    const preferredContextId = routing.preferredContextId;
    const windowMode = params.windowMode ?? envWindowMode;
    let id = generateId();
    let ensureUsed = false;
    let semanticRetryUsed = false;
    const ensureBridge = async () => {
        const remainingSeconds = Math.ceil((deadlineAt - Date.now()) / 1000);
        await ensureBrowserBridgeReady({
            timeoutSeconds: Math.max(1, Math.min(DEFAULT_BROWSER_CONNECT_TIMEOUT, remainingSeconds)),
            contextId,
            verbose: false,
        });
    };
    for (let attempt = 1; attempt <= TRANSPORT_MAX_ATTEMPTS; attempt++) {
        if (attempt > 1 && Date.now() >= deadlineAt) {
            throw new BrowserCommandError('Browser command deadline exhausted across transport retries.', COMMAND_RESULT_UNKNOWN_CODE, COMMAND_RESULT_UNKNOWN_HINT);
        }
        const remainingMs = Math.max(1000, deadlineAt - Date.now());
        const run = action === 'lease-release' ? undefined : getDaemonRunContext();
        const command = {
            id,
            action,
            ...params,
            timeout: timeoutSeconds,
            deadlineAt,
            ...(contextId && { contextId }),
            ...(preferredContextId && { preferredContextId }),
            ...(windowMode && { windowMode }),
            ...(run && {
                runId: run.runId,
                command: run.command,
                access: run.access,
                pid: process.pid,
            }),
        };
        try {
            const res = await requestDaemon('/command', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(command),
                timeout: remainingMs + HTTP_TIMEOUT_MARGIN_MS,
            });
            const result = (await res.json());
            if (result.ok)
                return result;
            if (res.status === 409 && result.code === 'session_busy' && result.holder) {
                throw new SessionBusyError(result.holder);
            }
            if (result.errorCode && UNKNOWN_OUTCOME_CODES.has(result.errorCode)) {
                throw new BrowserCommandError(result.error ?? 'Browser command result is unknown', result.errorCode, result.errorHint);
            }
            const isDuplicateCommandId = res.status === 409
                && !result.errorCode
                && (result.error ?? '').includes('Duplicate command id');
            if (isDuplicateCommandId) {
                throw new BrowserCommandError('Daemon already has this command id pending; the command may already be running.', COMMAND_RESULT_UNKNOWN_CODE, COMMAND_RESULT_UNKNOWN_HINT);
            }
            if (isPreDispatchError(result.errorCode) && !ensureUsed) {
                ensureUsed = true;
                await ensureBridge();
                continue;
            }
            const advice = classifyBrowserError(new BrowserCommandError(result.error ?? '', result.errorCode));
            if (advice.kind === 'extension-transient' && !semanticRetryUsed) {
                semanticRetryUsed = true;
                id = generateId();
                await sleep(advice.delayMs);
                continue;
            }
            throw new BrowserCommandError(result.error ?? 'Daemon command failed', result.errorCode, result.errorHint);
        }
        catch (err) {
            if (err instanceof BrowserCommandError || err instanceof BrowserConnectError)
                throw err;
            if (err instanceof Error && err.name === 'AbortError') {
                throw new BrowserCommandError('Browser command timed out client-side; the page may still have applied it.', COMMAND_RESULT_UNKNOWN_CODE, COMMAND_RESULT_UNKNOWN_HINT);
            }
            if (err instanceof TypeError) {
                await ensureBridge();
                if (isPreConnectFetchError(err))
                    continue;
                throw new BrowserCommandError('Connection to the daemon was lost mid-command; it may have already been applied.', COMMAND_RESULT_UNKNOWN_CODE, COMMAND_RESULT_UNKNOWN_HINT);
            }
            throw err;
        }
    }
    throw new BrowserCommandError('sendCommand: max attempts exhausted', 'max_attempts_exhausted');
}
/**
 * Send a command to the daemon and return the result data.
 */
export async function sendCommand(action, params = {}) {
    const result = await sendCommandRaw(action, params);
    return result.data;
}
/**
 * Like sendCommand, but returns both data and page identity (targetId).
 * Use this for page-scoped commands where the caller needs the page identity.
 */
export async function sendCommandFull(action, params = {}) {
    const result = await sendCommandRaw(action, params);
    return { data: result.data, page: result.page };
}
export async function releaseSiteSessionLease(runId) {
    const command = {
        id: generateId(),
        action: 'lease-release',
        runId,
    };
    await requestDaemon('/command', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(command),
        timeout: 2_000,
    }).catch(() => undefined);
}
export async function bindTab(session, opts = {}) {
    return sendCommand('bind', { session, surface: 'browser', ...opts });
}
